applications/product/groovyScripts/product/promo/ProductPromoActionServices.groovy

```diff
@@ -173,18 +183,19 @@ def productGWP() {
             Iterator<String> optionProductIdTempIter = optionProductIds.iterator()
             productId = optionProductIdTempIter.next()
             optionProductIdTempIter.remove()
-            product = from("Product").where("productId", productId).cache().queryOne()
+            product = from('Product').where('productId', productId).cache().queryOne()
         }
         if (!product) {
             // no product found to add as GWP, just return
-            return
+            return [*:result, actionResultInfo: parameters.actionResultInfo]
```

Comment:
Why is return different?

Reply:
IIRW This one took me some time to guess, it was revealed by ReturnsNullInsteadOfEmptyCollection, thus when returning [] was causing issue. After investigation, it is a service, and in that regard, it should return success or error, not nothing.


> Fix codenarc ReturnsNullInsteadOfEmptyCollection rules : Consider returning a zero length collection rather than null. It is often a better design to return a length zero collection rather than a null reference to indicate that there are no results (i.e., an empty list of results). This way, no explicit check for null is needed by clients of the method.

Reply:
OK, thanks for the info